Sources/GRPC/AsyncAwaitSupport/GRPCAsyncServerCallContext.swift

```diff
@@ -15,114 +15,104 @@
+  public func withUserInfo<Result: Sendable>(
+    _ body: @Sendable @escaping (UserInfo) throws -> Result
+  ) async throws -> Result {
+    return try await self.contextProvider.withUserInfo(body)
   }
 
-  private var _initialResponseMetadata: HPACKHeaders = [:]
-
-  /// Metadata to return at the end of the RPC.
+  /// Modify the `UserInfo` dictionary which is shared with the interceptor contexts for this RPC.
   ///
-  /// If this is required it should be updated before returning from the handler.
-  public var trailingResponseMetadata: HPACKHeaders {
-    get { self.lock.withLock {
-      return self._trailingResponseMetadata
-    } }
-    set { self.lock.withLock {
-      self._trailingResponseMetadata = newValue
-    } }
+  /// - Important: While `UserInfo` has value-semantics, this function accesses a reference
+  ///   wrapped `UserInfo`. The contexts passed to interceptors provide the same reference. As such
+  ///   this may be used as a mechanism to pass information between interceptors and service
+  ///   providers.
+  public func withMutableUserInfo<Result: Sendable>(
+    _ modify: @Sendable @escaping (inout UserInfo) -> Result
+  ) async throws -> Result {
+    return try await self.contextProvider.withMutableUserInfo(modify)
   }
 
-  private var _trailingResponseMetadata: HPACKHeaders = [:]
-
   @inlinable
   internal init(
     headers: HPACKHeaders,
     logger: Logger,
-    userInfoRef: Ref<UserInfo>
+    contextProvider: AsyncServerCallContextProvider
   ) {
-    self.requestMetadata = headers
-    self.userInfoRef = userInfoRef
-    self._logger = logger
+    self.request = Request(headers: headers, logger: logger)
+    self.contextProvider = contextProvider
+  }
+}
+
+@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+extension GRPCAsyncServerCallContext {
+  public struct Request: Sendable {
+    /// The request headers received from the client at the start of the RPC.
+    public var headers: HPACKHeaders
+
+    /// A logger.
+    public var logger: Logger
+
+    @usableFromInline
+    init(headers: HPACKHeaders, logger: Logger) {
+      self.headers = headers
+      self.logger = logger
+    }
+  }
+
+  public struct Response {
+    private let contextProvider: AsyncServerCallContextProvider
+
+    /// Set the metadata to return at the start of the RPC.
+    ///
+    /// - Important: If this is required it should be updated _before_ the first response is sent
+    ///   via the response stream writer. Updates must not be made after the first response has
+    ///   been sent.
+    public func setHeaders(_ headers: HPACKHeaders) async throws {
+      try await self.contextProvider.setResponseHeaders(headers)
+    }
+
+    /// Set the metadata to return at the end of the RPC.
+    ///
+    /// If this is required it must be updated before returning from the handler.
+    public func setTrailers(_ trailers: HPACKHeaders) async throws {
+      try await self.contextProvider.setResponseTrailers(trailers)
+    }
```

Comment:
The new API only allows to set headers/trailers. Should we expose APIs that allow to mutate the currently set headers?

Reply:
Good question. I thought about this a bit and I don't think we should provide it.

They are empty by default and the caller _knows_ what the value is: it can only have been set by them so I'm not sure it's actually useful to have (any update could be achieved by just setting the value a bit later).